linera-views/src/dynamo_db.rs

```diff
@@ -887,9 +823,208 @@ impl DynamoDbClientInternal {
+}
+
+#[doc(hidden)]
+#[allow(clippy::type_complexity)]
+pub struct DynamoDbKeyValueIteratorOwned {
+    prefix_len: usize,
+    pos: usize,
+    iters: Vec<
+        std::iter::Flatten<
+            std::option::IntoIter<Vec<HashMap<std::string::String, AttributeValue>>>,
+        >,
+    >,
+}
+
+impl Iterator for DynamoDbKeyValueIteratorOwned {
+    type Item = Result<(Vec<u8>, Vec<u8>), DynamoDbContextError>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let result = self.iters[self.pos].next();
+        match result {
+            None => {
+                if self.pos == self.iters.len() - 1 {
+                    return None;
+                }
+                self.pos += 1;
+                self.iters[self.pos]
+                    .next()
+                    .map(|mut x| extract_key_value_owned(self.prefix_len, &mut x))
+            }
+            Some(mut result) => Some(extract_key_value_owned(self.prefix_len, &mut result)),
+        }
+    }
+}
+
+impl KeyValueIterable<DynamoDbContextError> for DynamoDbKeyValues {
+    type Iterator<'a> = DynamoDbKeyValueIterator<'a> where Self: 'a;
+    type IteratorOwned = DynamoDbKeyValueIteratorOwned;
+
+    fn iterator(&self) -> Self::Iterator<'_> {
+        let pos = 0;
+        let mut iters = Vec::new();
+        for response in &self.result_queries.responses {
+            let iter = response.items.iter().flatten();
+            iters.push(iter);
+        }
+        DynamoDbKeyValueIterator {
+            prefix_len: self.result_queries.prefix_len,
+            pos,
+            iters,
+        }
+    }
+
+    fn into_iterator_owned(self) -> Self::IteratorOwned {
+        let pos = 0;
+        let mut iters = Vec::new();
+        for response in self.result_queries.responses.into_iter() {
+            let iter = response.items.into_iter().flatten();
+            iters.push(iter);
+        }
+        DynamoDbKeyValueIteratorOwned {
+            prefix_len: self.result_queries.prefix_len,
+            pos,
+            iters,
+        }
+    }
+}
+
+impl DynamoDbClientInternal {
+    async fn get_list_responses(
+        &self,
+        attribute: &str,
+        key_prefix: &[u8],
+    ) -> Result<QueryResults, DynamoDbContextError> {
+        if key_prefix.is_empty() {
+            return Err(DynamoDbContextError::ZeroLengthKeyPrefix);
+        }
+        if key_prefix.len() > MAX_KEY_BYTES {
+            return Err(DynamoDbContextError::KeyPrefixTooLong);
+        }
+        let mut responses = Vec::new();
+        let mut start_key = None;
+        loop {
+            let response = self
+                .get_query_output(attribute, key_prefix, start_key)
+                .await?;
+            let last_evaluated = response.last_evaluated_key.clone();
+            responses.push(response);
+            match last_evaluated {
+                None => {
+                    break;
+                }
+                Some(value) => {
+                    start_key = Some(value);
```

Comment:
Was the "last evaluated" key included in the previous response? And will it be included in the next response, too, i.e. will this duplicate the entry in the output?

Reply:
No. We are using the strategy indicated in https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Query.Pagination.html